cmd/root.go

```diff
@@ -423,3 +437,91 @@ func startHTTPServer(opts *common.Options) {
 	http.Handle("/metrics", promhttp.Handler())
 	http.ListenAndServe(opts.HTTPBindAddr, nil)
 }
+
+func proxyExchangeApis(server *grpc.Server) {
+	common.Log.Info("Proxying exchange APIs to light clients")
+
+	addTargetHandler := func(t tunnel.Target) error {
+		common.Log.WithFields(logrus.Fields{"target": t}).Warn("client attempted to register target")
+		return fmt.Errorf("client targets are not permitted")
+	}
+
+	// Allowlist of APIs that light clients are permitted to connect to.
+	// We have one target per API, because the light client needs to know the domain name
+	// to use for their TLS connection.
+	apis := make(map[tunnel.Target]string)
+	apis[tunnel.Target{ID: "binanceApi", Type: "HTTPS"}] = "api.binance.com:443"
+	apis[tunnel.Target{ID: "coinbaseApi", Type: "HTTPS"}] = "api.exchange.coinbase.com:443"
+	apis[tunnel.Target{ID: "geminiApi", Type: "HTTPS"}] = "api.gemini.com:443"
+
+	registerHandler := func(ss tunnel.ServerSession) error {
+		common.Log.WithFields(logrus.Fields{"addr": ss.Addr, "target": ss.Target}).Info("session requested")
+		for target := range apis {
+			if ss.Target.ID == target.ID {
+				return nil
+			}
+		}
+		return fmt.Errorf("target not allowed: %s", ss.Target.ID)
+	}
+
+	// Handler for a proxy connection.
+	sessionHandler := func(ss tunnel.ServerSession, rwc io.ReadWriteCloser) error {
+		common.Log.WithFields(logrus.Fields{"addr": ss.Addr, "target": ss.Target}).Info("new session")
+
+		var dialAddr string
+		for target, addr := range apis {
+			if ss.Target.ID == target.ID && ss.Target.Type == target.Type {
+				dialAddr = addr
+				break
+			}
+		}
+		if len(dialAddr) == 0 {
+			return fmt.Errorf("no matching dial port found for target: %s|%s", ss.Target.ID, ss.Target.Type)
+		}
+
+		// TODO: Enforce per-client (`ss.Addr`) rate limit.
```

Comment:
We can probably use https://pkg.go.dev/golang.org/x/time/rate for this. What we want is:
- A request rate limiter per exchange, that we configure with their respective public API limits.
- A session rate limiter per client, that we configure with some low rate that is sufficient for the kinds of requests that honest light clients will be making ("infrequent" queries of market tickers).
  - We can probably initially assume that we have one session per request, but HTTP connection pooling on the light client side may break this assumption.
  - Once we have request-response detection, we can change this to (or add) a request rate limiter per client.
- Some mechanism for fairly distributing per-exchange request rate limit tokens amongst connected clients (subject to the per-client rate limiter).